tests/js/common/aql/aql-insert-multiple-cluster.js

```diff
@@ -0,0 +1,724 @@
+
+// //////////////////////////////////////////////////////////////////////////////
+// / DISCLAIMER
+// /
+// / Copyright 2023 ArangoDB GmbH, Cologne, Germany
+// /
+// / Licensed under the Apache License, Version 2.0 (the "License")
+// / you may not use this file except in compliance with the License.
+// / You may obtain a copy of the License at
+// /
+// /     http://www.apache.org/licenses/LICENSE-2.0
+// /
+// / Unless required by applicable law or agreed to in writing, software
+// / distributed under the License is distributed on an "AS IS" BASIS,
+// / WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// / See the License for the specific language governing permissions and
+// / limitations under the License.
+// /
+// / Copyright holder is triAGENS GmbH, Cologne, Germany
+// /
+// / @author Julia Puget
+// //////////////////////////////////////////////////////////////////////////////
+
+const jsunity = require("jsunity");
+const arangodb = require("@arangodb");
+const internal = require('internal');
+const ERRORS = require("@arangodb").errors;
+const isServer = require("@arangodb").isServer;
+const isEnterprise = require("internal").isEnterprise;
+const deriveTestSuite = require('@arangodb/test-helper').deriveTestSuite;
+const cn = "UnitTestsCollection";
+const db = arangodb.db;
+const numDocs = 10000;
+const ruleName = "optimize-cluster-multiple-document-operations";
+
+if (!isServer) {
+  global.AQL_EXPLAIN = (query, bindVars, options) => {
+    let res = arango.POST("/_api/explain", {query, bindVars, options});
+    return res;
+  };
+}
+      
+const assertRuleIsUsed = (query, bind = {}, options = {}) => {
+  let res = AQL_EXPLAIN(query, bind, options);
+  assertNotEqual(-1, res.plan.rules.indexOf(ruleName));
+  const nodes = res.plan.nodes.map((n) => n.type);
+  assertNotEqual(-1, nodes.indexOf('MultipleRemoteModificationNode'));
+};
+
+const assertRuleIsNotUsed = (query, bind = {}, options = {}) => {
+  let res = AQL_EXPLAIN(query, bind, options);
+  assertEqual(-1, res.plan.rules.indexOf(ruleName));
+  const nodes = res.plan.nodes.map((n) => n.type);
+  assertEqual(-1, nodes.indexOf('MultipleRemoteModificationNode'));
+};
+
+function InsertMultipleDocumentsSuite(params) {
+  'use strict';
+  const {numberOfShards, replicationFactor} = params;
+  return {
+    setUp: function () {
+      db._drop(cn);
+      db._create(cn, {numberOfShards, replicationFactor});
+    },
+
+    tearDown: function () {
+      db._drop(cn);
+    },
+    
+    testFromEnumerateListInvalidInputsNoArray: function () {
+      const queries = [
+        `LET docs = MERGE({}, {}) FOR doc IN docs INSERT doc INTO ${cn}`,
+        `LET docs = NOOPT({}) FOR doc IN docs INSERT doc INTO ${cn}`,
+        `LET docs = {} FOR doc IN docs INSERT doc INTO ${cn}`,
+        `FOR doc IN {} INSERT doc INTO ${cn}`,
+        `FOR doc IN NOOPT({}) INSERT doc INTO ${cn}`,
+      ];
```

Comment:
I potentially found a disallowed case, so please add a test for the following (and decide if it is allowed or not)
Please run this test with all other optimizer rules disabled, just to avoid changing of node-ordering somewhere.


```
LET illegalList = (FOR x IN ${cn} LIMIT 5 RETURN x)
LET confusePattern = SLEEP(1)
FOR d IN illegalList
  INSERT d INTO ${cn}
```

As far as I understood the optimizer rule pattern, it expects:
Any ConstCalc, followed by Enumerate, followed by Insert
Where the enumerate cannot be part of a loop.
I guess with the intention that the ConstCalc is the Input for the Enumerate.

However in my test suggestion the ConstCalc Sleep (something that does not use an input variable) could potentially trigger the Pattern, and allow more or less arbitrary complex query before, which my be okayish, but later I see that subqueries are tried to be excluded.

We probably get all of the queries used here to pass using the same trick

Reply:
Ok but this one doesn't fail, just would not trigger the optimizer rule, so I'll put it somewhere else in the test

Reply:
_All_ depdendencies of the EnumerateList node must either be a const calc or the singleton node. So your query does not match the pattern, and we should already have a test case for that.

Reply:
Sorry you are totally right. :+1: